HeterogeneousCore/AlpakaCore/README.md

```diff
@@ -0,0 +1,403 @@
+# Alpaka algorithms and modules in CMSSW
+
+## Introduction
+
+This page documents the Alpaka integration within CMSSW. For more information about Alpaka itself see the [Alpaka documentation](https://alpaka.readthedocs.io/en/latest/).
+
+### Compilation model
+
+The code in `Package/SubPackage/{interface,src,plugins,test}/alpaka` is compiled once for each enabled Alpaka backend. The `ALPAKA_ACCELERATOR_NAMESPACE` macro is substituted with a concrete, backend-specific namespace name in order to guarantee different symbol names for all backends, that allows for `cmsRun` to dynamically load any set of the backend libraries.
+
+The source files with `.dev.cc` suffix are compiled with the backend-specific device compiler. The other `.cc` source files are compiled with the host compiler.
+
+The `BuildFile.xml` must contain `<flags ALPAKA_BACKENDS="1"/>` to enable the behavior described above.
+
+## Overall guidelines
+
+* Minimize explicit blocking synchronization calls
+  * Avoid `alpaka::wait()`, non-cached memory buffer allocations
+* If you can, use `global::EDProducer` base class
+  * If you need per-stream storage
+    * For few objects consider using [`edm::StreamCache<T>`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_StreamCacheT) with the global module, or
+    * Use `stream::EDProducer`
+  * If you need to transfer some data back to host, use `stream::SynchronizingEDroducer`
+* All code using `ALPAKA_ACCELERATOR_NAMESPACE` should be placed in `Package/SubPackage/{interface,src,plugins,test}/alpaka` directory
+  * Alpaka-dependent code that uses templates instead of the namespace macro can be placed in `Package/SubPackage/interface` directory
+* All source files (not headers) using Alpaka device code (such as kernel call, functions called by kernels) must have a suffic `.dev.cc`, and be placed in the aforementioned `alpaka` subdirectory
+* Any code that `#include`s a header from the framework or from the `HeterogeneousCore/AlpakaCore`(?) must be separated from the Alpaka device code, and have the usual `.cc` suffix.
+  * A few framework headers are allowed to be used in `.dev.cc` files, e.g. `FWCore/Utilities/interface/Exception.h` or `FWCore/MessageLogger/interface/MessageLogger.h` (?)
```

Comment:
In the leading order I'd think all code in `HeterogeneousCore/AlpakaCore` is about the interaction with the framework, and therefore would not be allowed to be used in `.dev.cc` files.

In the next-to-leading order, maybe some individual headers could be useful in `.dev.cc`, like an ability to throw `cms::Exception` and use MessageLogger, and with limited set of headers the maintenance burden of keeping them usable also by device compilers would be tolerable?

Reply:
- `FWCore/Utilities/interface/CMSUnrollLoop.h` is OK as it explicitly supports CUDA and ROCm (we could move or copy these macros into Alpaka itself, but even then it would not be a problem if somebody uses this file);
  - other macro-only headers are also fine (`FWCore/Utilities/interface/stringize.h`, _etc._);
  - `FWCore/Utilities/interface/Exception.h` should be OK, as it does not depend on other packages, and as long as the exception are used only in host code, not in device code;
  - `FWCore/MessageLogger` might be OK, too; though my preference would be to keep the messages in the `.cc` files and have only device code and simple wrappers in `.dev.cc` files.

Reply:
By the way, `EventCache.h` and `QueueCache.h` should be in `.dev.cc` files (even though I don't think we have a use for them in user code); I was thinking to eventually push a reworked version directly into Alpaka.